syntax = "proto2";
package ofs.c2m;

message IpAddress {
	required string host = 1;
	required int32 port = 2;
}

message NodeInfo {
	required int32 cpu = 1;
	required int32 rss = 2;
	required int32 vss = 3;
	required int32 disk = 4;
	required bool fault = 5;
}

message Register {
	required int32 id = 1;
	required IpAddress outpost = 2;
	required IpAddress harbor = 3;
	required int32 rack = 4;
	required int32 dc = 5;
	required NodeInfo node = 6;
	required string extend = 7;
}

message Neighbor {
	required int32 id = 1;
	required IpAddress harbor = 2;
}

message NeighborNotify {
	required Neighbor neighbor = 1;
}

message BlockStatus {
	required int64 id = 1;
	required int64 version = 2;
	required int32 size = 3;
	required bool fault = 4;
}

message Heartbeat {
	required int32 id = 1;
	required NodeInfo node = 2;
}

message UpdataBlock {
	required int32 id = 1;
	required BlockStatus block = 2;
}

message ReportBlock {
	required int32 id = 1;
	repeated BlockStatus blocks = 2;
}

message RecoverBlock {
	required int64 blockid = 1;
	required int64 version = 2;
	required int64 lease = 3;
	repeated int32 copyTo = 4;
}

message CleanBlock {
	required int64 blockid = 1;
}

message CleanComplete {
	required int32 id = 1;
	required int64 blockid = 2;
}

message NeighborGossip {
	required Neighbor neighbor = 1;
}

message WriteNotify {
	required int64 blockid = 1;
	required int64 version = 2;
	required int64 newversion = 3;
	required int32 offset = 4;
	required bytes data = 5;
}

message AppendNotify {
	required int64 blockid = 1;
	required int64 version = 2;
	required int64 newversion = 3;
	required bytes data = 4;
}

message ResizeBlock {
	required int64 blockid = 1;
	required int64 version = 2;
	required int32 size = 3;
}

message RecoverBlockData {
	required int64 blockid = 1;
	required int64 version = 2;
	required int32 offset = 3;
	required bytes data = 4;
}

message RecoverBlockComplete {
	required int64 blockid = 1;
	required int64 version = 2;
}
